integration/querydsl/blaze-persistence-querydsl-expressions/src/main/java/com/blazebit/persistence/querydsl/BlazeCriteriaBuilderRenderer.java

```diff
@@ -0,0 +1,1308 @@
+                                    BeanPath<?> beanPath = new BeanPath<Object>(attribute.getElementClass(), pathExpression, attributeName);
+                                    setExpressionSubqueries(beanPath, new BlazeCriteriaBuilderRenderer.Function<String, Object>() {
+                                        @Override
+                                        public Object apply(String expression) {
+                                            return bindBuilder.select(expression);
+                                        }
+                                    }, new BlazeCriteriaBuilderRenderer.Function<String, MultipleSubqueryInitiator<?>>() {
+                                        @Override
+                                        public MultipleSubqueryInitiator<?> apply(String expression) {
+                                            return bindBuilder.selectSubqueries(expression);
+                                        }
+                                    });
+                                }
+                            }
+                        } else {
+                            for (int i = 0; i < projection.size(); i++) {
+                                Expression<?> projExpression = projection.get(i);
+                                Path<?> alias = null;
+                                Path<?> cteEntityPath = null;
+
+                                if (projExpression instanceof Operation) {
+                                    Operation<?> projOperation = (Operation<?>) projExpression;
+                                    if (projOperation.getOperator() == BIND) {
+                                        alias = (Path<?>) projOperation.getArg(1);
+                                        cteEntityPath = alias.getRoot();
+                                        projExpression = projOperation.getArg(0);
+                                    }
+                                }
+                                if (alias == null && cteAliases != null) {
+                                    alias = cteAliases.get(i);
+                                }
+
+                                if (alias != null) {
+                                    String aliasString = relativePathString(cteEntityPath, alias);
+                                    final SelectBuilder<?> bindBuilder = selectBaseCriteriaBuilder.bind(aliasString);
+                                    setExpressionSubqueries(projExpression, new BlazeCriteriaBuilderRenderer.Function<String, Object>() {
+                                        @Override
+                                        public Object apply(String expression) {
+                                            return bindBuilder.select(expression);
+                                        }
+                                    }, new BlazeCriteriaBuilderRenderer.Function<String, MultipleSubqueryInitiator<?>>() {
```

Comment:
Originally I used method references here, so the code used to be `setExpressionSubqueries(beanPath, bindBuilder::select, bindBuilder::selectSubqueries)`.

I didn't bother to create inner classes because most of them are only used once (i.e. I have these anonymous classes as well for JoinOnBuilder, WhereBuilder, Selectbuilder, HavingBuilder). This one is indeed used twice, so I can extract them. However, after the Java 8 migration these cases that can be optimized using method references will be harder to detect again. These simple anonymous classes will be detected with ease by the IDE, and I think are better than having the inner classes.

WDYT?

Reply:
I think you can make these implementations stateless singletons if you use the `BiFunction` approach which is less memory intensive.